### PR TITLE
Fix notification-section markAsRead

### PR DESCRIPTION
**Pull request by mcbouslog**
Staging branch URL:

I hope fixes #5301, as currently notifications still not marked as "read" (`delivered: true`) after clicking to related source.

Describe your changes.
- binds `markAsRead`
- refactor `markAsRead`, including filtering on `subscription_id`, though imperfect, I think best option that applies to all sources (i.e. comment, moderation, data request, etc.)
- [x] tests

# Required Manual Testing

- [ ] Does the non-logged in home page render correctly?
- [ ] Does the logged in home page render correctly?
- [ ] Does the projects page render correctly?
- [ ] Can you load project home pages?
- [ ] Can you load the classification page?
- [ ] Can you submit a classification?
- [ ] Does talk load correctly?
- [ ] Can you post a talk comment?

# Review Checklist

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `rm -rf node_modules/ && npm install` and app works as expected?
- [ ] If the component is in coffeescript, is it converted to ES6? Is it free of eslint errors? Is the conversion its own commit?
- [ ] Are the tests passing locally and on Travis?

# Optional

- [ ] Have you replaced any `ChangeListener` or `PromiseRenderer` components with code that updates component state?
- [ ] If changes are made to the classifier, does the dev classifier still work?
- [ ] Have you [resized and compressed](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/image-optimization) any image you've added?
- [ ] Have you added in [flow type annotations](https://flowtype.org/docs/type-annotations.html)?
- [ ] Have you followed the [Springer guidelines for commit messages](https://github.com/springernature/frontend-playbook/blob/master/git/git.md#commit-messages)?

### CODE REVIEW

File: app/pages/notifications/notification-section.jsx [hunk 1]
```diff
@@ -143,13 +144,13 @@ export default class NotificationSection extends Component {
...
-    relatedNotifications.push(readNotification);
-    const readRequests = relatedNotifications
+    const relatedNotifications = notifications.filter(notification => (
+      notification.subscription_id === readNotification.subscription_id
+        && notification.delivered === false
+    ));
+    const promises = relatedNotifications
       .map(relatedNotification => relatedNotification.update({ delivered: true }).save());
-    Promise.all(readRequests).then(() => notificationsCounter.update(user));
+    Promise.all(promises).then(() => notificationsCounter.update(user));
```
Comment (eatyourgreens) [inline app/pages/notifications/notification-section.jsx]: Are the promises in `promises` guaranteed to resolve? `Promise.all` will reject if any of them reject and there's no catch block to handle that case. I'm not familiar with the notifications code, but my guess is you want to update the notifications counter after all the API requests complete, even if some failed.

File: app/pages/notifications/notification-section.jsx [hunk 2]
```diff
@@ -143,13 +144,13 @@ export default class NotificationSection extends Component {
...
-    const relatedNotifications = notifications.filter(
-      notification => notification.source_id === readNotification.source_id
-    );
-    relatedNotifications.push(readNotification);
-    const readRequests = relatedNotifications
+    const relatedNotifications = notifications.filter(notification => (
+      notification.subscription_id === readNotification.subscription_id
+        && notification.delivered === false
+    ));
+    const promises = relatedNotifications
```
Comment (eatyourgreens) [inline app/pages/notifications/notification-section.jsx]: `promises` doesn't tell us what this variable is doing. Is there a better name that could be used here?

File: app/pages/notifications/notification-section.jsx (hunk 1 above)
Comment (eatyourgreens) [reply app/pages/notifications/notification-section.jsx]: The [MDN example of fail-fast behaviour](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all#Promise.all_fail-fast_behaviour) suggests you could add catch blocks to the individual requests too.

File: app/pages/notifications/notification-section.jsx [hunk 3]
```diff
@@ -143,13 +144,13 @@ export default class NotificationSection extends Component {
     const { notificationsCounter } = this.context;
     const { user } = this.props;
     const { notifications } = this.state;
-    const relatedNotifications = notifications.filter(
-      notification => notification.source_id === readNotification.source_id
-    );
-    relatedNotifications.push(readNotification);
-    const readRequests = relatedNotifications
+    const relatedNotifications = notifications.filter(notification => (
+      notification.subscription_id === readNotification.subscription_id
```
Comment (camallen) [inline app/pages/notifications/notification-section.jsx]: I think this needs to incorporate the `source_id` and `source_type` as well. I'm not sure the `subscription_id` matches across different users. 

I'm not entirely sure which attributes you need to 'relate' the notifications together but my testing resulted in the above attributes. I think you need to handle @ mentions, comments and other forms of notification events.

**Comment by eatyourgreens** [comment]
I think `NotificationsCounter` does update the counts, it's just hard to follow because it uses context.
https://github.com/zooniverse/Panoptes-Front-End/blob/cf0e6264aa067a53f9cd326ceba4e73a933e4e16/app/talk/lib/notifications-link.cjsx#L21-L26

**Comment by camallen** [comment]
Thanks @eatyourgreens context + callback on the `notifications-counter` was easy to miss.

**Comment by eatyourgreens** [comment]
React is designed specifically to avoid using event listeners, so it's easy to miss how these counts work. Depending on how much effort we want to put into this, the unread count and associated actions could move out of app state and into a redux store.

**Comment by mcbouslog** [comment]
Sharing [this gist](https://gist.github.com/mcbouslog/37f9c33b1ba12497e7d0b8a184f7765b) with a comment, reply, mention, moderation report and subject note (also a comment) used to attempt to determine "related notifications."

File: app/pages/notifications/notification-section.jsx (hunk 3 above)
Comment (mcbouslog) [reply app/pages/notifications/notification-section.jsx]: Based on [this gist](https://gist.github.com/mcbouslog/37f9c33b1ba12497e7d0b8a184f7765b) I think comments, replies and mentions in the same discussion should be marked as read if the user goes to said discussion, otherwise I think other notifications shouldn't be grouped (moderation reports, data requests) but marked as read individually. This is just my opinion though, of course open to input.

File: app/pages/notifications/notification-section.jsx [hunk 4]
```diff
@@ -143,13 +144,22 @@ export default class NotificationSection extends Component {
...
-    );
-    relatedNotifications.push(readNotification);
-    const readRequests = relatedNotifications
-      .map(relatedNotification => relatedNotification.update({ delivered: true }).save());
-    Promise.all(readRequests).then(() => notificationsCounter.update(user));
+    const relatedNotifications = notifications.filter(notification => (
+      notification.delivered === false
+        && (notification.id === readNotification.id
+          || notification.source.discussion_id === readNotification.source.discussion_id)
+    ));
```
Comment (mcbouslog) [inline app/pages/notifications/notification-section.jsx]: `notification.id === readNotification.id` captures notifications for moderation, data requests, whereas `notification.source.discussion_id === readNotification.source.discussion_id` captures comments (which includes replies and mentions) related to a clicked comment notification

Comment (camallen) [reply app/pages/notifications/notification-section.jsx]: LGTM

**Comment by mcbouslog** [comment]
Related to #4727
